homeassistant/components/toon/__init__.py

```diff
@@ -1,289 +1,155 @@
-from homeassistant.helpers.entity import Entity
-from homeassistant.helpers.event import async_track_time_interval
-from homeassistant.helpers.typing import ConfigType, HomeAssistantType
-
-from . import config_flow  # noqa: F401
-from .const import (
-    CONF_DISPLAY,
-    CONF_TENANT,
-    DATA_TOON,
-    DATA_TOON_CLIENT,
-    DATA_TOON_CONFIG,
-    DATA_TOON_UPDATED,
-    DEFAULT_SCAN_INTERVAL,
-    DOMAIN,
+from homeassistant.helpers.config_entry_oauth2_flow import (
+    OAuth2Session,
+    async_get_config_entry_implementation,
 )
+from homeassistant.helpers.typing import ConfigType
+
+from .const import CONF_AGREEMENT_ID, DOMAIN
+from .coordinator import ToonDataUpdateCoordinator
+from .oauth2 import register_oauth2_implementations
+
+ENTITY_COMPONENTS = {
+    BINARY_SENSOR_DOMAIN,
+    CLIMATE_DOMAIN,
+    SENSOR_DOMAIN,
+    SWITCH_DOMAIN,
+}
 
 _LOGGER = logging.getLogger(__name__)
 
 # Validation of the user's configuration
 CONFIG_SCHEMA = vol.Schema(
     {
-        DOMAIN: vol.Schema(
-            {
-                vol.Required(CONF_CLIENT_ID): cv.string,
-                vol.Required(CONF_CLIENT_SECRET): cv.string,
-                vol.Required(
-                    CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
-                ): vol.All(cv.time_period, cv.positive_timedelta),
-            }
+        DOMAIN: vol.All(
+            cv.deprecated(CONF_SCAN_INTERVAL),
+            vol.Schema(
+                {
+                    vol.Required(CONF_CLIENT_ID): cv.string,
+                    vol.Required(CONF_CLIENT_SECRET): cv.string,
+                }
+            ),
         )
     },
     extra=vol.ALLOW_EXTRA,
 )
 
-SERVICE_SCHEMA = vol.Schema({vol.Optional(CONF_DISPLAY): cv.string})
-
 
-async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up the Toon components."""
     if DOMAIN not in config:
         return True
 
-    conf = config[DOMAIN]
+    register_oauth2_implementations(
+        hass, config[DOMAIN][CONF_CLIENT_ID], config[DOMAIN][CONF_CLIENT_SECRET]
+    )
 
-    # Store config to be used during entry setup
-    hass.data[DATA_TOON_CONFIG] = conf
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(DOMAIN, context={"source": SOURCE_IMPORT})
+    )
 
     return True
 
 
-async def async_setup_entry(hass: HomeAssistantType, entry: ConfigType) -> bool:
-    """Set up Toon from a config entry."""
+async def async_migrate_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Handle migration of a previous version config entry."""
+    if entry.version == 1:
+        # There is no usable data in version 1 anymore.
+        # The integration switched to OAuth and because of this, uses
+        # different unique identifiers as well.
+        # Force this by removing the existing entry and trigger a new flow.
+        await hass.config_entries.async_remove(entry.entry_id)
```

Comment:
Let's remove it once the import was successful ? Otherwise one might restart in between and no longer be prompted to authenticate